live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/adapters/storage/PreSignedUrlRequesterTest.java

```diff
@@ -0,0 +1,202 @@
+
+import static org.alfresco.hxi_connector.live_ingester.adapters.storage.PreSignedUrlRequester.STORAGE_LOCATION_PROPERTY;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.camel.CamelContext;
+import org.apache.camel.Exchange;
+import org.apache.camel.Message;
+import org.apache.camel.Processor;
+import org.apache.camel.ProducerTemplate;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import org.alfresco.hxi_connector.live_ingester.domain.exception.LiveIngesterRuntimeException;
+import org.alfresco.hxi_connector.live_ingester.domain.ports.storage.StorageLocationRequest;
+
+@ExtendWith(MockitoExtension.class)
+class PreSignedUrlRequesterTest
+{
+    private static final int STATUS_CODE_201 = 201;
+    private static final String STORAGE_LOCATION = "http://dummy-url";
+    private static final String RESPONSE_BODY = "{\"" + STORAGE_LOCATION_PROPERTY + "\": \"" + STORAGE_LOCATION + "\"}";
+
+    @Mock
+    CamelContext camelContextMock;
+    @Mock
+    ProducerTemplate producerTemplateMock;
+    @Mock
+    Exchange exchangeMock;
+    @Mock(strictness = Mock.Strictness.LENIENT)
+    Message messageMock;
+    @Mock
+    ObjectMapper objectMapperMock;
+
+    @InjectMocks
+    PreSignedUrlRequester preSignedUrlRequester;
+
+    @BeforeEach
+    void setUp()
+    {
+        given(camelContextMock.createProducerTemplate()).willReturn(producerTemplateMock);
+        given(producerTemplateMock.send(any(String.class), any(Processor.class))).willReturn(exchangeMock);
+        given(exchangeMock.getMessage()).willReturn(messageMock);
+        given(messageMock.getHeader(any(String.class), any(Class.class))).willReturn(STATUS_CODE_201);
+        given(messageMock.getBody(any())).willReturn(RESPONSE_BODY);
+    }
+
+    @Test
+    void testRequestStorageLocation() throws IOException
+    {
+        // given
+        StorageLocationRequest request = createStorageLocationRequestMock();
+        given(objectMapperMock.readValue(any(String.class), any(TypeReference.class))).willReturn(Map.of(STORAGE_LOCATION_PROPERTY, STORAGE_LOCATION));
+
+        // when
+        URL url = preSignedUrlRequester.requestStorageLocation(request);
+
+        // then
+        then(messageMock).should().getHeader(Exchange.HTTP_RESPONSE_CODE, Integer.class);
+        then(messageMock).should().getBody(String.class);
+        then(objectMapperMock).should().readValue(eq(RESPONSE_BODY), anyTypeReferenceOfMap());
+        assertThat(url).asString().isEqualTo(STORAGE_LOCATION);
+    }
```

Comment:
Let's think what this test is supposed to test - imo it would be:
- After I will call `requestStorageLocation()` method
- and I will receive valid response from some service
- I would like to get `url` service send me

but what we are testing now is:
- After I will call `requestStorageLocation()` method
- and something will call ObjectMapper.readValue() with some than return map of properties
- and on response we received from hxi `getHeader()` method should be called
- and on response we received from hxi `getBody()` method should be called
- and `readValue()` method should be called on objectMapper
- and I would like to get `url` service send me

which imo contains way to much implementation details. What I would propose to do instead is:

```
  @Mock
    CamelContext camelContextMock;
    @Mock
    ProducerTemplate producerTemplateMock;
    @Spy
    ObjectMapper objectMapper = new ObjectMapper();

    @InjectMocks
    PreSignedUrlRequester preSignedUrlRequester;

    @Test
    void testRequestStorageLocation()
    {
        // given
        StorageLocationRequest request = new StorageLocationRequest("node-ref", "content/type");
        httpClientShouldRespondWith(RESPONSE_BODY, STATUS_CODE_201);

        // when
        URL url = preSignedUrlRequester.requestStorageLocation(request);

        // then
        assertThat(url).asString().isEqualTo(STORAGE_LOCATION);
    }
```

```
    void httpClientShouldRespondWith(String body, int statusCode) {
        Exchange exchangeMock = mock();
        Message messageMock = mock();

        given(camelContextMock.createProducerTemplate()).willReturn(producerTemplateMock);
        given(producerTemplateMock.send(any(String.class), any(Processor.class))).willReturn(exchangeMock);
        given(exchangeMock.getMessage()).willReturn(messageMock);
        given(messageMock.getHeader(Exchange.HTTP_RESPONSE_CODE, Integer.class)).willReturn(statusCode);
        given(messageMock.getBody(any())).willReturn(body);
    }
```
This contains much less code tightly coupled with internal implementation and is much more readable (and I would propose to do the same for all other test methods as well)

Reply:
I let Camel to handle more logic, so unit tests were rewritten.